test_agent/rust/src/utils.rs

```diff
@@ -0,0 +1,552 @@
+        if!(_entity_version_major == 0)
+        {
+        _entity.version_major = Some(_entity_version_major)
+        };
+        if!(_entity_version_minor == 0)
+        {
+        _entity.version_minor = Some(_entity_version_minor)
+        };
+        _entity.special_fields = _entity_special_fields;
+      //  dbg!("_entity: {:?}", _entity);
+        let ___entity = MessageField(Some(Box::new(_entity)));
+
+        let _resource_name = match value
+            .get("resource")
+            .and_then(|resource| resource.get("name"))
+        {
+            Some(_resource_name) => _resource_name.as_str().expect("issue in name"),
+            None => "default",
+        };
+        let _resource_instance = match value
+            .get("resource")
+            .and_then(|resource| resource.get("instance"))
+        {
+            Some(_resource_instance) => _resource_instance.as_str().map(|s| s.to_owned()),
+            None => Some(String::from("default")),
+        };
+        let _resource_message = match value
+            .get("resource")
+            .and_then(|resource| resource.get("message"))
+        {
+            Some(_resource_message) => _resource_message.as_str().map(|s| s.to_owned()),
+            None => Some(String::from("default")),
+        };
+        let _resource_id = match value
+            .get("resource")
+            .and_then(|resource| resource.get("id"))
+        {
+            Some(_resource_id) => _resource_id
+                .clone()
+                .as_str()
+                .expect("not a string")
+                .parse::<u32>()
+                .expect("issue in converting to u32"),
+            None => 0,
+        };
+        let _resource_special_fields = SpecialFields::default();
+        let mut _resource = UResource::new();
+        _resource.name = _resource_name.to_owned();
+        _resource.instance = _resource_instance;
+        _resource.message = _resource_message;
+        if!(_resource_id == 0)
+        {
+        _resource.id = Some(_resource_id)
+    };
+
+        let ___resource = MessageField(Some(Box::new(_resource)));
+        let _special_fields = SpecialFields::default();
+    let mut _uuri:UUri = UUri::new();
+    if!(_authority_name.clone() == Some("default") && _authority_number_id.clone() == vec![0] && _authority_number_ip.clone() == vec![0]){
+        dbg!("authority is not default");
+        _uuri.authority = MessageField(Some(Box::new(_authority)));
+    }
+    _uuri.entity = ___entity;
+    _uuri.resource = ___resource;
+
+
+        Ok(WrapperUUri(_uuri))
+    }
+}
+#[derive(Default)]
+pub struct WrapperUAttribute(pub UAttributes);
+impl<'de> Deserialize<'de> for WrapperUAttribute {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let value: Value = Deserialize::deserialize(deserializer)?;
+        // Conversion function from string to enum variant
+        fn from_str_comstatus(s: &str) -> UCode {
```

Comment:
For those items involving `UCode`, I think you can work with the existing methods available from the protobuf crate without having to write your own, e.g.

```rust
let ucode = UCode::from_str("UNKNOWN");
println!("ucode: {ucode:?}");

let ucode_desc = ucode.unwrap().descriptor();
let ucode_str = ucode_desc.name();
println!("ucode_str: {ucode_str}");
```

seems to work and gets me:
```
running 1 test
ucode: Some(UNKNOWN)
ucode_str: UNKNOWN
test ustatus::tests::test_convert_ucode_to_str ... ok
```

Reply:
corrected.